From 4647cb048b46763bb808b55f2ae2737cabc4d4a9 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Tue, 23 Aug 2022 12:11:22 +0300
Subject: [PATCH] [SW-99461] gaudi3: add persist section in rom

Boot info infrastructure is introduced in gaudi3 ROM. It is
the first platform to use this infrastructure in ROM code.
As a general practice, as the code would be frozen, a
contingency flow should be in place.

Boot info is currenlty used in ROM boot flow to read ppboot,
tpm and fake fuse image descriptors. Hence as a contingency
plan these image descriptors and relavent data needs to be
stored in an alternate memory which is persistent in nature,
as re-dma of the image descriptors is done in cold-reset and
fw update, all the other resets need to re-use the stored
data.

Hence we store the arcboot_rom_global data structure in the
initial 3KB of DCCM. This memory region would be carved out
such that it would persist through out the boot flow.

Change-Id: Ib5311baae738e8ad9d3927774f7db3a8e35ab427
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 945b188..2ea038c 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -16,7 +16,7 @@
 #include "boot_desc.h"
 #include <common.h>
 
-struct arcboot_global_s arcboot_rom_global = {
+struct arcboot_global_s __attribute__((section (".rom_persistent_section"))) arcboot_rom_global = {
 	.config_fuse = 0,
 	.config2_fuse = 0,
 	.recovery_flag = 0,
diff --git a/arch/gaudi2/arc/arcboot_rom.lds b/arch/gaudi2/arc/arcboot_rom.lds
index 55904dd..80af301 100644
--- a/arch/gaudi2/arc/arcboot_rom.lds
+++ b/arch/gaudi2/arc/arcboot_rom.lds
@@ -46,6 +46,7 @@
 
 	.data : AT ( LOADADDR(.text) + SIZEOF(.text) ) {
 		_data = .;
+		KEEP(*(.rom_persistent_section))
 		*(.data)
 	} > REGION_DATA
 
diff --git a/arch/gaudi3/arc/arcboot_rom.lds b/arch/gaudi3/arc/arcboot_rom.lds
index 2c3d74d..47263d3 100644
--- a/arch/gaudi3/arc/arcboot_rom.lds
+++ b/arch/gaudi3/arc/arcboot_rom.lds
@@ -46,6 +46,8 @@
 
 	.data : AT ( LOADADDR(.text) + SIZEOF(.text) ) {
 		_data = .;
+	    KEEP(*(.rom_persistent_section))
+	    . = _data + 3k; /* Reserve 3KB memory for persistent data section */
 		*(.data)
 	} > REGION_DATA
 
diff --git a/arch/greco/arc_elf32/arcboot_rom.lds b/arch/greco/arc_elf32/arcboot_rom.lds
index cce0716..1cbe290 100644
--- a/arch/greco/arc_elf32/arcboot_rom.lds
+++ b/arch/greco/arc_elf32/arcboot_rom.lds
@@ -48,6 +48,7 @@
 
 	.data : AT ( LOADADDR(.text) + SIZEOF(.text) ) {
 		_data = .;
+		 KEEP(*(.rom_persistent_section))
 		*(.data)
 	} > REGION_DATA
 
